implement Levenberg-Marquardt first draft

amir-saadat opened the pull request:
Hi Mathieu, this is the first draft of Levenberg-Marquardt implementation. Please let me know your feedback. I also had some issues with the implicit differentiation unit test and it would be great if you can take a look and let me know if you encountered any issues. I am not passing on implicit_diff=True of test_double_exponential_fit_implicit_diff unit test.

mblondel reviewed (commented):
Thanks @amir-saadat! I made a pass on the code, still need to make a pass on the tests.

amir-saadat commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,313 @@
+    verbose: whether to print error on every iteration or not.
+      Warning: verbose=True will automatically disable jit.
+    jit: whether to JIT-compile the bisection loop (default: "auto").
+    unroll: whether to unroll the bisection loop (default: "auto").
```
Good point thanks.

amir-saadat commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,313 @@
+  """Named tuple containing state information."""
+  iter_num: int
+  damping_coeff: float
+  nu: float
```
For sure, increase_factor works just fine.

amir-saadat commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,313 @@
+
+    # found = jnp.linalg.norm(grad, jnp.inf) <= gtol
+    jtj_diag = self._jtj_diag_op(init_params, *args, **kwargs)
+    damping_coeff = self.damping_parameter * jnp.max(jtj_diag)
```
For sure. Updated `damping_coeff` to `damping_factor`.

amir-saadat commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,313 @@
+
+    return LevenbergMarquardtState(iter_num=jnp.asarray(0),
+                                   damping_coeff=damping_coeff,
+                                   nu=nu,
```
Done.

amir-saadat replied:
Done. thanks.

amir-saadat replied:
Done. thanks.

amir-saadat commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,313 @@
+
+  damping_parameter: float = 1e-6
+  xtol: float = 1e-8
+  gtol: float = 1e-8
```
I am not sure if that would be possible, if we would like to keep the algorithm as is. The algorithm for this implementation suggests stopping either if gradients are small or the delta_params is small. If we have only one tol, we wouldn't be able to distinguish between the two, right? Does that address your question?

The drawback of the current usage is that tol needs to be zero in order to preserve the stopping criteria of the algorithm, at least this was the only way i could think of that doesn't require changing the `_cond_fun` for the `state.error` in the `base.py`.

mblondel replied:
So far, all solvers had a single stopping criterion, typically the l2 norm of  optimality_fun (the gradient in your case). Since optimality_fun is also used for implicit_diff, I think this makes sense and would be my preference for simplicity and consistency.

If you really want to keep , we would indeed need a way to override `_cond_fun` in base.py, as I'm not a fan of using `tol = 0`.

amir-saadat replied:
I chose to do something in between: I override the `_cond_fun` now but give the user the freedom to choose the `stop_criteria`. WDYT?

mblondel replied:
This works for me. In the future, we may want to provide a public API to do that. This implementation will be part of JAXopt so it's fine to override a private method but for external implementations, it's not ok.

amir-saadat commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,313 @@
+
+      gain_ratio = (loss_curr - loss_next) / gain_ratio_denom
+
+      def gain_ratio_test_true_func(cond_state):
```
I think that is a great idea. I extracted two functions, the second one being called in the first one, and the first one is called within the `update` function. Please double check if consistent with your thoughts.

amir-saadat commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,313 @@
+
+      def gain_ratio_test_false_func(cond_state):
+
+        params, error, grad_error, damping_coeff, nu, _, _ = cond_state
```
Done. Thanks for the suggestion.

amir-saadat commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,313 @@
+            jtj_reg_matvec, state.gradient, init=state.delta)
+    elif self.solver == 'cholesky':
+      delta_params = solve_cholesky(jtj_reg_matvec, state.gradient)
+    elif self.solver == 'inv':
```
Done. thanks for the suggestion.

amir-saadat commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,313 @@
+    loss_curr = state.loss
+
+    def jtj_reg_matvec(vec):
+      """The (J^T.J + damping_coeff*I) matrix vector product operator."""
```
Done. thanks for the suggestion.

amir-saadat commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,313 @@
+      """The (J^T.J + damping_coeff*I) matrix vector product operator."""
+      return self._jtj_reg_op(params, state.damping_coeff, vec, *args, **kwargs)
+
+    if self.solver == 'cg':
```
Done. Thanks.

amir-saadat commented on tests/gauss_newton_test.py:
```diff
@@ -83,7 +83,7 @@ def test_enzyme_reaction_parameter_fit(self):
 
 
-  @parameterized.product(implicit_diff=[True, False])
+  @parameterized.product(implicit_diff=[False, False])
```
I haven't forgot this :). It is intentionally set as False as well, cause `True` doesn't pass the test. I basically tried what I did for Gauss-Newton, while the example here has many more data points (the number of x and y values are 40+). It would be great if you can take a look and see if I am using the implicit differentiation correctly. Thanks.

mblondel replied:
But you didn't change anything in `gauss_newton.py` so why should the tests stop working?

amir-saadat replied:
Oh my bad. I meant to avoid implicit differentiation in the LM file and I must have looked at the wrong file when making the change. Thank you for pointing that out.

amir-saadat commented:
@mblondel thanks for your comments/suggestions. I think I have addressed most of your first round comments. I am still having issues with the implicit differentiation unit test. It would be great if you can take a look. Thank you

mblondel commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,356 @@
+    residual, aux = self._fun_with_aux(init_params, *args, **kwargs)
+    gradient = self._jt_op(init_params, residual, *args, **kwargs)
+
+    grad_diff = tree_sub(tree_inf_norm(gradient), self.gtol)
```
`tree_inf_norm` returns a scalar value so normally you can write

```suggestion
    grad_diff = tree_inf_norm(gradient) - self.gtol
```

You have a bunch of those in the code.

amir-saadat replied:
Done. thanks for catching.

mblondel commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,356 @@
+    # for the value of dparams.
+
+    tree_mul_term = tree_mul(self.xtol,
+                             tree_add(tree_l2_norm(params), self.xtol))
```
Same here and below. Normally `tree_l2_norm` returns a scalar so you don't need to use tree utils on top.

amir-saadat replied:
Done. thanks for catching.

mblondel commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,356 @@
+
+    def jtj_reg_matvec(vec):
+      """The (J^T.J + damping_factor*I) matrix vector product operator."""
+      return self._jtj_reg_op(params, state.damping_factor, vec, *args,
```
Since you added the `ridge` option directly to the solvers I think this should be
```suggestion
      return self._jtj_op(params, state.damping_factor, vec, *args,
```
And in this case `def _jtj_reg_op` can be deleted.

amir-saadat replied:
Actually `jtj_reg_matvec ` can be deleted now, since all of the solvers are using `ridge` and I kept the `jtj_reg_matvec` for the custom solver. If we are assuming custom solver will have the `ridge` option, we can just remove this definition.

mblondel commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,356 @@
+    return self._jtj_op(params, vec, *args, **
+                        kwargs) + lam * jnp.dot(jnp.eye(len(params)), vec)
+
+  def _jtj_diag_op(self, params, *args, **kwargs):
```
Unfortunately, I think this is just as expensive as computing the full Jacobian once for all and store it. I see that this information is used for updating the damping factor. Does it mean that the Levenberg-Marquardt algorithm is not "matvec" friendly? Maybe @shoyer has some thoughts on this.

amir-saadat replied:
This is one way to initialize the damping factor, but it is pretty typical. Although, this is only once during the optimization.

mblondel replied:
If it's done only once in the initialization, then that's fine! Sorry I thought it was done in `update`.

amir-saadat replied:
no worries, yeah it is just for `init_state`.

mblondel commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,356 @@
+
+    if isinstance(self.solver, Callable):
+      delta_params = self.solver(
+          jtj_reg_matvec, state.gradient, init=state.delta)
```
the `ridge` option is missing

amir-saadat replied:
Done. thanks.

mblondel commented on tests/levenberg_marquardt_test.py:
```diff
@@ -0,0 +1,149 @@
+  # provided by K. Madsen & H. B. Nielsen in the book "Introduction to
+  # Optimization and Data Fitting" Example 6.10-. page 124. Data taken from
+  # immoptibox "A MATLAB TOOLBOX FOR OPTIMIZATION AND DATA FITTING" H.B. Nielsen
+  @parameterized.product(solver=['cg', 'bicgstab', 'gmres', 'cholesky', 'inv'])
```
Maybe we should only accept 'cholesky', 'inv' and Callable. This way we don't have to test all possible solvers (we try to keep tests short when possible).

amir-saadat replied:
Sure :).

mblondel commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,372 @@
+      derived for updating the coefficients using Gauss-Newton method. Please
+      see section 3.2. of K. Madsen et al. in the book "Methods for nonlinear
+      least squares problems" for more information.
+    stop_criteria: The criteria to use for the convergence of the while loop.
```
I would call it `stopping_criterion` (`criteria` is plural)

amir-saadat replied:
Done.

mblondel commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,372 @@
+  tol: float = 0.0
+
+  damping_parameter: float = 1e-6
+  stop_criteria: Literal['grad-l2', 'madsen-nielsen'] = 'madsen-nielsen'
```
I think `grad-l2-norm` would be slightly more explicit.

Also I would use `grad-l2-norm` by default. This is consistent with all other solvers in JAXopt and uses `tol`.

amir-saadat replied:
Done.

mblondel commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,372 @@
+
+  damping_parameter: float = 1e-6
+  stop_criteria: Literal['grad-l2', 'madsen-nielsen'] = 'madsen-nielsen'
+  tol: float = 0.0
```
If `grad-l2-norm` is the default, then `tol` should be non-zero.

mblondel commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,372 @@
+  damping_parameter: float = 1e-6
+  stop_criteria: Literal['grad-l2', 'madsen-nielsen'] = 'madsen-nielsen'
+  tol: float = 0.0
+  xtol: float = 1e-8
```
`1e-8` may be very difficult to reach for float32 precision, which is the default precision we operate with. How about `1e-3`?

mblondel commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,372 @@
+        jtrpp = self._jt_op(params, rpp, *args, **kwargs)
+        delta_params += 0.5*self.solver(matvec, jtrpp, ridge=state.damping_factor)
+    elif self.solver == 'cg':
+      delta_params = solve_cg(
```
Any reason to keep cg explicitly? Can't we support it via Callable? Especially since I see you don't test this branch anymore.

amir-saadat replied:
Done. Thanks.

mblondel commented on tests/levenberg_marquardt_test.py:
```diff
@@ -0,0 +1,163 @@
+        optimize_info.params, onp.array([4, -4, 4, 5], dtype=float), atol=1e-3)
+    # Convergence after 62 iterations.
+    if geodesic:
+      self.assertAllClose(optimize_info.state.iter_num, 20)
```
Nice!

amir-saadat replied:
Yes! it is very exciting.

mblondel reviewed (changes requested):
Thanks a lot @amir-saadat for taking all my comments into account!

Could you also add `LevenvergMarquardt` to [api.rst](https://github.com/google/jaxopt/blob/main/docs/api.rst) and add a [changelog](https://github.com/google/jaxopt/blob/main/docs/api.rst) entry.

amir-saadat commented:
Thanks @mblondel for great comments. I hope they are addressed now.

amir-saadat commented on jaxopt/_src/levenberg_marquardt.py:
```diff
@@ -0,0 +1,362 @@
+  xtol: float = 1e-3
+  gtol: float = 1e-3
+
+  solver: Union[Literal['cholesky', 'inv'], Callable] = 'cholesky'
```
Sure, I changed it to `solve_cg` I think that would make more sense since the matrix is symmetric PSD.

amir-saadat commented:
Thanks @mblondel for the comments. I updated api.rst, and changelog.rst. Please double check if I got it right. I added the changes for a new version v.3.02. The previous unit test was very slow (several min) and I decided to use another unit test that is similar to GaussNewton example. That is working fine now.

mblondel commented on tests/levenberg_marquardt_test.py:
```diff
@@ -0,0 +1,182 @@
+from unittest import SkipTest
+
+import jax
+from jax import test_util as jtu
```
Remove

amir-saadat replied:
Removed.

mblondel commented on tests/levenberg_marquardt_test.py:
```diff
@@ -0,0 +1,182 @@
+    self.assertArraysAllClose(jac_num, jac_custom, atol=1e-2)
+
+if __name__ == '__main__':
+  absltest.main(testLoader=jtu.JaxTestLoader())
```
```suggestion
  absltest.main()
```

amir-saadat replied:
Applied. thanks.

mblondel commented on docs/changelog.rst:
```diff
@@ -7,6 +7,17 @@ Main branch
 ~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+Version 0.3.2.
+--------------
```
You can put your modification in "Main branch" for now. I am not sure what version number we should use yet.

amir-saadat replied:
sounds good. thanks.
